CAF-3877: Initial implementation of workflow worker

mmaBel opened the pull request.

Commit:
CAF-3877: Initial implementation of workflow worker.

mmaBel commented:
The workflow transformation logic was previously reviewed under;
https://github.com/CAFDataProcessing/data-processing-service/pull/21

Commit:
CAF-3877: Update workflow xslt to recognize the Chained Action type.

dermot-hardy commented on README.md:
```diff
@@ -1,5 +1,25 @@
+
+### workfow-to-js-transformer
+
+This project is a Java library to retrieve a workflow and transform it to a JavaScript representation of the action logic. It can be found in [workfow-to-js-transformer](workfow-to-js-transformer).
```
"workflow" misspelt several times, including in the link, which is broken.

mmaBel replied:
Fixed.

dermot-hardy commented on pom.xml:
```diff
@@ -0,0 +1,117 @@
+        <dockerHubOrganization>cafdataprocessing</dockerHubOrganization>
+        <dockerCafDataProcessingOrg>${dockerImagePrefix}${dockerHubOrganization}${dockerOrgSeperator}</dockerCafDataProcessingOrg>
+        <dockerProjectVersion>${dockerVersionSeperator}${project.version}</dockerProjectVersion>
+        <processingServiceVersion>1.4.0-SNAPSHOT</processingServiceVersion>
```
I don't think this property (i.e. `processingServiceVersion`) is useful since there is no value you could move it to that would work for a release version - i.e. for the Docker image usage you would require the `<major>.<minor>` form whereas for the Maven usage you would require `<major>.<minor>.<patch>-<buildNumber>` form.

mmaBel replied:
Re-using it for the container won't work yeah but the property is useful I think as it keep the version numbers that change centrally located under that properties section for ease of updating.

Removed its use from the container property.

dermot-hardy commented on pom.xml:
```diff
@@ -0,0 +1,117 @@
+            <dependency>
+                <groupId>com.github.cafdataprocessing</groupId>
+                <artifactId>workflow-to-js-transformer</artifactId>
+                <version>${project.version}</version>
```
It's best to avoid using `${project.version}` for dependencies in any POM which might be used as a parent of another POM.  Just use the `1.0.0-SNAPSHOT` instead and let the build machine will update it.  It's unlikely to cause a real issue in this case since although this POM will be used as a parent it is unlikely to be used as a parent by any projects outside of this repository.

mmaBel replied:
OK

dermot-hardy commented on pom.xml:
```diff
@@ -0,0 +1,117 @@
+    <properties>
+        <caf.worker-document-framework.version>3.3.0-SNAPSHOT</caf.worker-document-framework.version>
+        <corePolicyVersion>1.1.0-SNAPSHOT</corePolicyVersion>
+        <corePolicyElasticsearchContainerName>cafinternal/prereleases:policy-admin-elasticsearch-${corePolicyVersion}</corePolicyElasticsearchContainerName>
```
If possible cut the length of this line and line 52 as they go over the 138-character guideline.

mmaBel replied:
Fixed.

dermot-hardy commented on worker-workflow-constants/src/main/java/com/github/cafdataprocessing/workflow/constants/WorkflowWorkerConstants.java:
```diff
@@ -0,0 +1,49 @@
+/**
+ * Constant values relating to the Workflow Worker
+ */
+public class WorkflowWorkerConstants {
```
Make this class `final` and give it a private default constructor to make sure that nobody tries to derive from it or instantiate it.

mmaBel replied:
OK

dermot-hardy commented on worker-workflow-constants/src/main/java/com/github/cafdataprocessing/workflow/constants/WorkflowWorkerConstants.java:
```diff
@@ -0,0 +1,49 @@
+    /**
+     * Error codes describing reason for a failure to process documents.
+     */
+    public static class ErrorCodes {
```
What is the intention of this module?  I assume it is for public consumption, to define constants which different workers in the workflow need to agree on?  If so, should these error codes be here?  Are they private to the worker?  I'm not sure.

mmaBel replied:
It's for constant values related specifically to this workflow worker (not other workers), such as the custom data. The caller is going to need to know the key value to set so they can pull in this library to do that (as the family task submitter does). The Error codes is a similar principle where someone may want to be able to reference the error code values and handle them appropriately if encountered.

dermot-hardy commented on worker-workflow-container/README.md:
```diff
@@ -0,0 +1,60 @@
+| threads       |  `CAF_WORKFLOW_WORKER_THREADS`          |      1       |
+|               |  `CAF_WORKER_THREADS`          |             |
+| maxBatchSize  |  `CAF_WORKER_MAX_BATCH_SIZE`          |             |
+| maxBatchTime  |  `CAF_WORKER_MAX_BATCH_TIME`          |             |
```
This isn't a bulk worker so `maxBatchSize` and `maxBatchTime` are not used.

mmaBel replied:
OK. Removed them from the configuration file also.

dermot-hardy commented on worker-workflow-container/README.md:
```diff
@@ -0,0 +1,60 @@
+
+| Property | Checked Environment Variables                        | Default               |
+|----------|------------------------------------------------------|-----------------------|
+| processingApiUrl   |  `CAF_WORKFLOW_WORKER_PROCESSING_API_URL`  |                       |
```
I think an example would be helpful with this one, just to show what a typical processing api endpoint looks like.

mmaBel replied:
OK. The compose file has an example also.

dermot-hardy commented on worker-workflow-container/src/main/config/cfg~caf~worker~DocumentWorkerConfiguration.js:
```diff
@@ -0,0 +1,30 @@
+    failureQueue: getenv("CAF_WORKER_FAILURE_QUEUE") || undefined,
+    threads: getenv("CAF_WORKFLOW_WORKER_THREADS") || getenv("CAF_WORKER_THREADS") || 1,
+    maxBatchSize: getenv("CAF_WORKER_MAX_BATCH_SIZE") || undefined,
+    maxBatchTime: getenv("CAF_WORKER_MAX_BATCH_TIME") || undefined,
```
The `maxBatchSize` and `maxBatchTime` lines are harmless so you can leave them if you want to but they would only become useful if we update the code to implement [the `BulkDocumentWorker` interface](https://cafdataprocessing.github.io/data-processing-service/pages/en-us/Creating_an_Action/apidocs/com/hpe/caf/worker/document/extensibility/BulkDocumentWorker.html).

mmaBel replied:
Removed them.

dermot-hardy commented on worker-workflow-container/worker.sh:
```diff
@@ -0,0 +1,43 @@
+
+# If the CAF_APPNAME and CAF_CONFIG_PATH environment variables are not set, then use the
+# JavaScript-encoded config files that are built into the container
+if [ -z "$CAF_APPNAME" ] && [ -z "$CAF_CONFIG_PATH" ];
```
The reason we do this `if` statement in a lot of workers is for backward compatibility, so that users could move from file-based config to env-based config on their own schedule.  For a new worker, I'm not sure that we really don't want to just set these four environment variables all the time, so that there is no option for file-based configuration.  Do you think we should keep the option of file-based configuration open for this worker for any reason?

mmaBel replied:
I don't see value in letting them control the decoder, appname and substitutor but I can see a use for being able to point the config path to another location if someone wanted to alter the logic to derive the config properties (be that for production or debug).

Edit: Thinking on it, if there's a need to use a custom config file they could just overwrite the /maven/config folder with new files. I'll remove the 'if' statement and always export those four environment variables.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/WorkflowWorkerConfiguration.java:
```diff
@@ -0,0 +1,50 @@
+        return workflowCachePeriod;
+    }
+
+    public void setProcessingApiUrl(String processingApiUrl) {
```
Use `final` for parameters whose value doesn't change in the implementation of the method.  (I could make this point in other places also but I'm not going to.)

mmaBel replied:
OK.

dermot-hardy commented on worker-workflow/README.md:
```diff
@@ -0,0 +1,36 @@
+
+The worker utilizes the following configuration properties. The environment variables to control these and defaults values can be seen in the container for the worker, [here](../worker-workflow-container).
+
+### processingApiUrl
```
I see there is a link that I can click on to get the environment variables, and I know why you're putting the config keys here and the environment variables in the container Readme, but from an external perspective it doesn't seem right to call out the config keys which is essentially internal and not the environment variable.  I think we should at least directly write what the corresponding environment variable is.

mmaBel replied:
I'd like to avoid duplicating that environment variable information. The keys are mentioned here just to give some description, how about if I move the key details into that configuration table on the container page and link to the configuration from here?

dermot-hardy replied:
OK let's do that then.

mmaBel replied:
Added 'description' column to the container config properties table and moved text there.

dermot-hardy commented on worker-workflow/README.md:
```diff
@@ -0,0 +1,36 @@
+
+The main places where the worker can fail are;
+
+- Configuration errors: these will manifest on startup. Check the logs for clues, and double check your configuration settings.
```
I'm not necessarily asking for this to be changed but the best way to deal with configuration errors would be to allow the worker to start up, but for the health check to immediately fail.  If the user directly calls the health check endpoint then in the Json they should see that it is a configuration error and what variable is faulty.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/CustomDataExtractor.java:
```diff
@@ -0,0 +1,113 @@
+/**
+ * Given a document this class returns specified properties from the document's custom data.
+ */
+class CustomDataExtractor
```
Mark this class and all the other classes in this module `final`.  Only don't mark a class `final` if you envisage that it might be useful to subclass from it.

mmaBel replied:
OK.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/CustomDataExtractor.java:
```diff
@@ -0,0 +1,113 @@
+            document.addFailure(WorkflowWorkerConstants.ErrorCodes.INVALID_CUSTOM_DATA,
+                    "No output partial reference value passed to worker in custom data.");
+            customDataValid = false;
+        }
```
I don't think it's right that `outputPartialReference` is mandatory.  Looking in the `worker-store-fs` code it explicitely says that it "may be null" --> https://github.com/WorkerFramework/worker-framework/blob/v2.3.1/worker-store-fs/src/main/java/com/hpe/caf/worker/datastore/fs/FileSystemDataStore.java#L229.  I think there may have been an issue in some workers which expected it but those are issues in those workers.

mmaBel replied:
OK. Changed this to only log that message at debug level and not add a failure if partial reference isn't passed. Updated the README doc to specify which are optional and required.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/CustomDataExtractor.java:
```diff
@@ -0,0 +1,113 @@
+                workflowId = extractedWorkflowId;
+            }
+        }
+        catch(NumberFormatException e) {
```
I know I said I wouldn't go on about `final` but I just want to point out that exception catches should also usually be `final`:

    catch(final NumberFormatException e) {
        ...
    }

mmaBel replied:
OK.

dermot-hardy commented on worker-workflow/pom.xml:
```diff
@@ -0,0 +1,76 @@
+                    <groupId>javax.ws.rs</groupId>
+                    <artifactId>jsr311-api</artifactId>
+                </exclusion>
+            </exclusions>
```
I suspect this exclusion should actually be in the `workflow-to-js-transformer` module as presumably the `jsr311-api` module is never required to matter how `workflow-to-js-transformer` is used?

mmaBel replied:
The processing client used in the transformer is pulling it in. The transformer module needs it to be there unless whoever is importing is already using a different version, as is the case with the worker.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/WorkflowWorkerFactory.java:
```diff
@@ -0,0 +1,49 @@
+ */
+public class WorkflowWorkerFactory implements DocumentWorkerFactory {
+    private final static Logger LOG = LoggerFactory.getLogger(WorkflowWorkerFactory.class);
+    private WorkflowWorkerConfiguration workflowWorkerConfiguration = null;
```
Why is this an instance variable rather than just a local variable in `createDocumentWorker()`?

mmaBel replied:
Removed as part of passing application to worker instead,

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/WorkflowWorkerFactory.java:
```diff
@@ -0,0 +1,49 @@
+            }
+        }
+        return new WorkflowWorker(workflowWorkerConfiguration, application.getService(DataStore.class));
+    }
```
This point is very arguable but I think it would be better if this class was grossly simplified and the entire implementation of this method was just `return new WorkflowWorker(application);`.  The `WorkflowWorker` constructor could take an `Application` object instead.

Either way I would be inclined to move your `workflowWorkerConfiguration` instantiation logic into a helper method.  Something like:

    private static WorkflowWorkerConfiguration getWorkflowWorkerConfiguration(final Application application)
    {
        try {
            return application.getService(ConfigurationSource.class)
                              .getConfiguration(WorkflowWorkerConfiguration.class);
        } catch (ConfigurationException e) {
            LOG.warn("Unable to load WorkflowWorkerConfiguration. "
                + "Specific Workflow Worker Worker configuration will not be passed to the worker.");
            return new WorkflowWorkerConfiguration();
        }
    }

mmaBel replied:
OK, changed to pass Application to worker and get config and datastore during instantiation. This was based on what we do for other document workers that have factories defined.
[Family Hashing](https://github.houston.softwaregrp.net/caf/worker-familyhashing/blob/develop/worker-familyhashing/src/main/java/com/hpe/caf/worker/WorkerFamilyHashingFactory.java) and [Lang Detect](https://github.com/CAFDataProcessing/worker-languagedetection/blob/develop/worker-languagedetection/src/main/java/com/hpe/caf/worker/languagedetection/LanguageDetectionWorkerFactory.java).

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/WorkflowWorker.java:
```diff
@@ -0,0 +1,182 @@
+        this.processingApiUrl = workflowWorkerConfiguration.getProcessingApiUrl();
+        final ApiClient apiClient = new ApiClient();
+        apiClient.setBasePath(this.processingApiUrl);
+        this.workflowAdminApi = new AdminApi(apiClient);
```
I think these three lines should be pulled out into a `getWorkflowAdminApi()` helper method which takes a `String` and returns an `AdminApi`.

mmaBel replied:
OK.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/WorkflowWorker.java:
```diff
@@ -0,0 +1,182 @@
+    private String storeWorkflow(String workflowJavaScript, String outputPartialReference)
+            throws DataStoreException
+    {
+        return dataStore.store(workflowJavaScript.getBytes(Charset.forName("UTF-8")), outputPartialReference);
```
Use `StandardCharsets.UTF_8` rather than `Charset.forName("UTF-8")`.

mmaBel replied:
OK.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/TransformedWorkflowCache.java:
```diff
@@ -0,0 +1,96 @@
+     * @return constructed cache key.
+     */
+    private static String buildCacheKey(final long workflowId, final String projectId) {
+        return Long.toString(workflowId) + projectId;
```
The key being built here is potentially ambiguous.  workflowId=4,projectId="3Par" will get the same key as workflowId=43,projectId="Par".

You could fix this by adding a non-numeric delimiter between the ids.  A more complex but ultimately better fix would be to introduce a dedicated class to use as the key of the cache instead of using a `String` - you'd need to implement `equals()` and `hashCode()` in the class but you'd never need to do string concatenation.

mmaBel replied:
Added dedicated class TransformedWorkflowCacheKey.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/TransformedWorkflowCache.java:
```diff
@@ -0,0 +1,96 @@
+ */
+public class TransformedWorkflowCache
+{
+    private static final ConcurrentHashMap<String, ReentrantLock> TRANSFORM_WORKFLOW_LOCKS = new ConcurrentHashMap<>();
```
Why is this `static`?  Why would you want to share `ReentrantLock`s across multiple caches?

mmaBel replied:
Originally the cache itself was static when it was in the factory. Missed this when moving its creation into the worker class. Removed static.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/TransformedWorkflowCache.java:
```diff
@@ -0,0 +1,96 @@
+     */
+    public void addTransformWorkflowResult(final long workflowId, final String projectId,
+                                           final TransformWorkflowResult transformWorkflowResult) {
+        workflowCache.asMap().putIfAbsent(buildCacheKey(workflowId, projectId), transformWorkflowResult);
```
Why not just use `workflowCache.put()`?  I.e. if we have a new result then wouldn't we want to use it regardless of the previous result?

mmaBel replied:
Can't recall reason to prefer that approach and can't think of any downside with using put so changed to that.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/WorkflowWorker.java:
```diff
@@ -0,0 +1,182 @@
+            finally {
+                transformWorkflowLock.unlock();
+            }
+        }
```
The logic here seems sound (it doesn't suffer Java's usual double-locking pattern issue - see [Double-checked locking: Clever, but broken](https://www.javaworld.com/article/2074979/java-concurrency/double-checked-locking--clever--but-broken.html) if you don't know what I'm talking about), but it's also overly complicated - you are using the Guava Cache object but you would get the extra functionality you're hand-coding with the `ConcurrentHashMap` if you used the Guava LoadingCache instead.  See [LoadingCache.html#get(K)](https://google.github.io/guava/releases/19.0/api/docs/com/google/common/cache/LoadingCache.html#get(K)).

mmaBel replied:
I had started with LoadingCache but couldn't think of a way to add  specific failures during transformation to a document if the loader was in charge of performing the transformation (it would only have the key information available to it). Will take another look at it and discuss it next year.

dermot-hardy replied:
The answer to that is with an exception.  If you can't load the cache element then you can throw an exception which you can catch back in the code where you have the document context.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/WorkflowWorker.java:
```diff
@@ -0,0 +1,182 @@
+            workflowJavaScript = WorkflowTransformer.retrieveAndTransformWorkflowToJavaScript(
+                    extractedProperties.getWorkflowId(), extractedProperties.getProjectId(),
+                    processingApiUrl);
+        } catch (ApiException | WorkflowTransformerException e) {
```
I'm not sure what these exceptions mean but if an `ApiException` means that the Processing API is down then a `DocumentWorkerTransientException` should be thrown in response rather than adding a failure to the document.  The reasons are that:
1. this document should be re-tried when it is back up
2. no other documents should be tried until it is back up, and throwing the transient exception will automatically stop the worker from accepting more tasks from the input queue until it becomes healthy again.

mmaBel replied:
ApiException can happen for multple reasons;
- occurs if there are issues in the request trying to be sent
- if the response received from API is considered unsuccessful e.g. a 500 error.

I can't recall if the API being down is an APIException (I think it is) but APIException doesn't necessarily imply a transient problem.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/WorkflowWorker.java:
```diff
@@ -0,0 +1,182 @@
+import com.hpe.caf.worker.document.exceptions.DocumentWorkerTransientException;
+import com.hpe.caf.worker.document.exceptions.PostProcessingFailedException;
+import com.hpe.caf.worker.document.extensibility.DocumentWorker;
+import com.hpe.caf.worker.document.model.*;
```
List the classes explicitly rather than using a star-import
http://checkstyle.sourceforge.net/config_imports.html#AvoidStarImport

mmaBel replied:
OK.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/WorkflowEvaluator.java:
```diff
@@ -0,0 +1,70 @@
+/**
+ * Evaluates a JavaScript workflow against a document, updating document based on matched actions in the workflow
+ */
+public class WorkflowEvaluator
```
Add a `private` empty default constructor to avoid this class being instantiated.

mmaBel replied:
OK

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/WorkflowEvaluator.java:
```diff
@@ -0,0 +1,70 @@
+            responseOptions.setCustomData(customData);
+        }
+        else {
+            responseOptions.getCustomData().put(DocumentWorkerConstants.POST_PROCESSING_SCRIPT_CUSTOM_DATA, workflowStorageReference);
```
This line is effectively a duplicate of line 61.  If you deal with the customData null potential first, by setting an empty map, then it wouldn't need to be duplicated.

mmaBel replied:
If customData is set it looks to be unmodifiable (on ResponseImpl) so I've changed this to copy all the existing customData values (if it is not null) to an empty map, add the post processing script and set the custom data on response options.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/WorkflowEvaluator.java:
```diff
@@ -0,0 +1,70 @@
+    {
+        final JavaScriptDocumentPostProcessor jsPostProcessor = new JavaScriptDocumentPostProcessor(workflowAsJavaScript);
+        jsPostProcessor.postProcessDocument(document);
+        setWorkflowPostProcessingScript(document, workflowStorageRef);
```
We don't need to do this right now but raise a Jira to add some functionality to the Document Worker Framework so that you don't need this dependency on `JavaScriptDocumentPostProcessor`.  Really the worker should only need to depend on `worker-document-interface`, but obviously you are having to take a dependency on `worker-document` itself, which is meant to be an internal module, in order to use this class.

mmaBel replied:
Raised, https://jira.autonomy.com/browse/CAF-3963.

Commit:
CAF-3878: Apply review changes.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/TransformedWorkflowCacheKey.java:
```diff
@@ -0,0 +1,47 @@
+        if(this == o) {
+            return true;
+        }
+        if(o == null || getClass() != o.getClass()) {
```
Using `if(o == null || getClass() != o.getClass())` is fine but because the class is `final` there is no possibility of any subclasses so you can use `if (!(o instanceof TransformedWorkflowCacheKey))` instead which I think is a little easier.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/TransformedWorkflowCacheKey.java:
```diff
@@ -0,0 +1,47 @@
+        }
+        TransformedWorkflowCacheKey cacheKeyToCheck = (TransformedWorkflowCacheKey) o;
+        return Objects.equals(this.projectId, cacheKeyToCheck.getProjectId()) &&
+                Objects.equals(this.workflowId, cacheKeyToCheck.getWorkflowId());
```
Using `Objects.equals(this.workflowId, cacheKeyToCheck.getWorkflowId())` will box both values.  You should just use `(this.workflowId == cacheKeyToCheck.workflowId)` instead.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/WorkflowWorker.java:
```diff
@@ -0,0 +1,206 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.charset.Charset;
```
Unused import

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/WorkflowWorker.java:
```diff
@@ -0,0 +1,206 @@
+            workflowJavaScript = WorkflowTransformer.retrieveAndTransformWorkflowToJavaScript(
+                    extractedProperties.getWorkflowId(), extractedProperties.getProjectId(),
+                    processingApiUrl);
+        } catch (final ApiException | WorkflowTransformerException e) {
```
So you're saying for some `ApiException`s we should throw a `DocumentWorkerTransientException` and for others we shouldn't.  How can we differentiate between these?

mmaBel replied:
Following discussion, the plan is that when an ApiException is thrown we will call the processing API health check. If that also throws an ApiException we will throw a transient exception. If it succeeds then we will retry the retrieve and transform once, adding a failure to the document if an ApiException occurs again.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/WorkflowWorkerFactory.java:
```diff
@@ -0,0 +1,37 @@
+import com.hpe.caf.worker.document.extensibility.DocumentWorkerFactory;
+import com.hpe.caf.worker.document.model.Application;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
There are unsued imports here now.

mmaBel replied:
Removed.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/WorkflowWorkerFactory.java:
```diff
@@ -0,0 +1,37 @@
+ * A factory to create workflow workers, passing them a configuration instance.
+ */
+public final class WorkflowWorkerFactory implements DocumentWorkerFactory {
+    private final static Logger LOG = LoggerFactory.getLogger(WorkflowWorkerFactory.class);
```
This logger is unused also.

mmaBel replied:
Removed.

dermot-hardy reviewed (changes requested):
I'd still like to see the `LoadingCache` change made as I think that will remove some complexity.

Commit:
CAF-3878: Apply review changes. Use loading cache instead of custom implementation when retrieving transformed workflow. Throw transient exception if a workflow transform appears to fail due to the processing API being unhealthy. Removing support for file based config. Moving configuration documentation around to centralize it.

mmaBel commented:
Moved to LoadingCache instead of custom implementation and added retry mechanism when APIException/UncheckedExecutionException occurs
- Type of exception varies depending on circumstance for failure
e.g. 
- Policy Database down - APIException
- Policy API down - ClientHandlerException (wrapping a SocketException)
- Processing API down - ClientHandlerException (wrapping an UnknownHostException)
(ClientHandlerException is a jersey exception, rather than tie to the dependencies of the processing client the UncheckedExecutionException and Exception types have been used instead).

Commit:
CAF-3877: Changes to transform retry logic following testing. Add CAF_WORKER_JAVA_OPTS environment variable to worker in line with other CAF workers.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/WorkflowWorker.java:
```diff
@@ -0,0 +1,203 @@
+            }
+            healthMonitor.reportUnhealthy("Processing API communication is unhealthy. Service dependencies:"
+                    +dependenciesStatusBuilder.toString());
+        } catch (final Exception e) {
```
You've changed from catching `ApiException` here to catching `Exception`.  That will cause you to catch `RuntimeException`s as well which you really shouldn't have to catch.

mmaBel replied:
Changed as part of update to throw a new WorkflowRetrievalException when the jersey code throws its runtime exception.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/WorkflowWorker.java:
```diff
@@ -0,0 +1,203 @@
+
+        // Get the specified workflow and transform it to a JavaScript representation
+        final TransformWorkflowResult transformWorkflowResult = transformWorkflow(extractedProperties, document);
+        if(transformWorkflowResult==null){
```
Is it ever valid for this to be `null`?  Let's discuss this branch when you get a chance.

mmaBel replied:
Discussed. It is valid for this to be null in  certain api, transformation and datastore exceptions.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/WorkflowWorker.java:
```diff
@@ -0,0 +1,203 @@
+                        throw new DocumentWorkerTransientException(
+                                "Unable to transform workflow. Processing API is unhealthy.");
+                    }
+                } catch (final Exception secondException) {
```
I don't like that this is catching `Exception`, but it's in the middle of the area you're currently changing so this might be going away.

mmaBel replied:
Yep this is no longer there.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/TransformedWorkflowCache.java:
```diff
@@ -0,0 +1,173 @@
+                .expireAfterWrite(workflowCachePeriod.get(ChronoUnit.SECONDS), TimeUnit.SECONDS)
+                .build(new CacheLoader<TransformedWorkflowCacheKey, TransformWorkflowResult>() {
+                    @Override
+                    public TransformWorkflowResult load(TransformedWorkflowCacheKey key)
```
Missing `final` specifier.

mmaBel replied:
Fixed.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/TransformedWorkflowCache.java:
```diff
@@ -0,0 +1,173 @@
+            // load
+            final Throwable cause = e.getCause();
+            if(cause instanceof ApiException ) {
+                LOG.error("Processing API exception when trying to retrieve workflow.", cause);
```
Are we double-logging here?  I see that you already logged this on the other side in the `transformWorkflow` function.  What is the correct level for it to be logged at?

mmaBel replied:
Yeah that's double-logging. Removed this and left it in the transform workflow as error level.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/TransformedWorkflowCache.java:
```diff
@@ -0,0 +1,173 @@
+     * @throws ApiException if a failure occurs during communication with processing API.
+     * @throws DataStoreException if a failure occurs storing transformed workflow.
+     * @throws WorkflowTransformerException if a failure occurs during transformation of the workflow.
+     * @throws
```
Extra `@throws` line.

mmaBel replied:
Removed.

dermot-hardy reviewed (changes requested):
A few minor changes as well as what we discussed.

Commit:
CAF-3877: Add new exception type indicating workflow retrieval failure to avoid catching UncheckedExecutionException when transforming workflow. Other minor code review comments.

Commit:
CAF-3877: Remove unused import.
